pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -1026,6 +1044,166 @@ func registriesConfigIgnition(templateDir string, controllerConfig *mcfgv1.Contr
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("error checking for %s config map: %w", configMapName, err)
+	}
+	// If the crio-default-container-runtime config map exists, the MC was already created, so skip creating it again.
+	if defaultContainerRuntimeCM != nil && !errors.IsNotFound(err) {
+		return nil
+	}
+
+	sel, err := metav1.LabelSelectorAsSelector(metav1.AddLabelToSelector(&metav1.LabelSelector{}, builtInLabelKey, ""))
+	if err != nil {
+		return err
+	}
+	// Find all the MachineConfigPools
+	mcpPoolsAll, err := ctrl.mcpLister.List(sel)
+	if err != nil {
+		return err
+	}
+
+	// Create the crio-default-container-runtime MC for all the available pools
+	for _, pool := range mcpPoolsAll {
+		if pool.Name != ctrlcommon.MachineConfigPoolMaster && pool.Name != ctrlcommon.MachineConfigPoolWorker {
+			continue
+		}
+		managedKey := getManagedKeyDefaultContainerRuntime(pool)
+		mc, err := ctrl.client.MachineconfigurationV1().MachineConfigs().Get(context.TODO(), managedKey, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			return fmt.Errorf("error checking for %s machine config: %w", managedKey, err)
+		}
+		// continue to the next MC if this already exists
+		if mc != nil && !errors.IsNotFound(err) {
+			continue
+		}
+
+		// Determine the container runtime for the final runtime configuration.
+		currentDefaultRuntime, err := ctrl.determineContainerRuntimeForFinalRuntimeConfig(pool)
+		if err != nil {
+			return fmt.Errorf("could not determine final runtime config for pool %s: %w", pool.Name, err)
+		}
+
+		if currentDefaultRuntime == mcfgv1.ContainerRuntimeDefaultRuntimeEmpty {
+			klog.Infof("default_runtime is not set in any existing MachineConfigs for pool %s, updating it", pool.Name)
+			// Update the default_runtime setting
+			tempIgnCfg := ctrlcommon.NewIgnConfig()
+			mc, err = ctrlcommon.MachineConfigFromIgnConfig(pool.Name, managedKey, tempIgnCfg)
+			if err != nil {
+				return fmt.Errorf("could not create crio-default-container-runtime MachineConfig from new Ignition config: %w", err)
+			}
+			rawRuntimeIgnition, err := json.Marshal(createNewIgnition(createDefaultContainerRuntimeFile()))
+			if err != nil {
+				return fmt.Errorf("error marshalling crio-default-container-runtime config ignition: %w", err)
+			}
+			mc.Spec.Config.Raw = rawRuntimeIgnition
+			// Create the crio-default-container-runtime MC
+			if err := retry.RetryOnConflict(updateBackoff, func() error {
+				_, err = ctrl.client.MachineconfigurationV1().MachineConfigs().Create(context.TODO(), mc, metav1.CreateOptions{})
+				return err
+			}); err != nil {
+				return fmt.Errorf("could not create MachineConfig for crio-default-container-runtime: %w", err)
+			}
+			klog.Infof("Applied default runtime MC %v on MachineConfigPool %v", managedKey, pool.Name)
+		} else {
+			klog.Infof("default_runtime is already set in existing MachineConfigs for pool %s, skipping MachineConfig creation", pool.Name)
+			continue
```

Comment:
Just to check something here, what we're saying is that if the configmap doesn't exist (the MCO never created a default runtime), but there is a set runtime already via other containerruntimeconfigs, we're not going to try to create a default containerruntimeconfig here, but we will try to create the configmap. Meaning that we essentially only try the "default creation" once upon the first upgrade to 4.17.

I guess the only time this would matter is if the user wants to delete their hard-set runtime sometime in the 4.17 timeframe, and in this case it would not try to set a default and just switch upon upgrading to 4.18. That's... probably the expected behaviour, but just wanted to check that.

Reply:
Yes, that's correct.